runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -73,96 +73,140 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
      */
-    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Range<Long> range, boolean waitForWrite) {
-        long startAddress = range.lowerEndpoint();
-        long endAddress = range.upperEndpoint();
+    @Override
+    @Nonnull
+    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout,
+                                         Range<Long> addresses,
+                                         boolean waitForWrite) {
+        long startAddress = addresses.lowerEndpoint();
+        long endAddress = addresses.upperEndpoint();
         int numUnits = runtimeLayout.getLayout().getSegmentLength(startAddress);
         log.trace("readRange[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
 
-        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(
-                runtimeLayout
-                        .getLogUnitClient(startAddress, numUnits - 1)
-                        .read(range)).getAddresses();
+        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(runtimeLayout
+                .getLogUnitClient(startAddress, numUnits - 1)
+                .readRange(addresses)).getAddresses();
+
+        return readOrHoleFill(runtimeLayout, logResult, startAddress, numUnits, waitForWrite);
+    }
+
+    private Map<Long, ILogData> readOrHoleFill(RuntimeLayout runtimeLayout,
+                                               Map<Long, LogData> logResult,
+                                               long startAddress,
+                                               int numUnits,
+                                               boolean waitForWrite) {
+        Map<Long, ILogData> returnResult = new TreeMap<>(logResult);
+
+        // If waiting not required for writer, hole fill directly
+        if (!waitForWrite) {
+            logResult.forEach((address, value) -> {
+                if (!isLogDataPresent(value)) {
+                    holeFill(runtimeLayout, address);
+                    returnResult.put(address, peek(runtimeLayout, address));
+                }
+            });
+            return returnResult;
+        }
 
         // In case of holes, use the standard backoff policy for hole fill for
-        // the first entry in the list. All subsequent holes in the list can
-        // be hole filled without waiting as we have already waited for the first
-        // hole.
-        boolean wait = !waitForWrite;
-        Map<Long, ILogData> returnResult = new TreeMap<>();
+        // the first entry in the list. All subsequent holes in the list can be
+        // hole filled without waiting as we have already waited for the first hole.
+        boolean wait = true;
         for (Map.Entry<Long, LogData> entry : logResult.entrySet()) {
             long address = entry.getKey();
             ILogData value = entry.getValue();
-            if (value == null || value.isEmpty()) {
-                if (!wait) {
+            if (!isLogDataPresent(value)) {
```

Comment:
Just use `ILogData::isEmpty`, lower layers shouldn't use null to signal empty positions.

Reply:
Done.